actors/builtin/miner/miner_actor.go

```diff
@@ -1097,315 +1371,332 @@ func (a Actor) OnDeferredCronEvent(rt Runtime, payload *CronEventPayload) *adt.E
-	if currDeadline.PeriodElapsed() {
-		// A cron event has not yet processed the previous proving period and established the next one.
-		// This is possible in the first non-empty epoch of a proving period if there was an empty tipset on the
-		// last epoch of the previous period.
-		// The period must be reset before processing missing-post faults.
-		rt.Abortf(exitcode.ErrIllegalState, "proving period at %d elapsed, next one not yet opened", currDeadline.PeriodStart)
+	// Record whether or not we _have_ early terminations now.
+	hasEarlyTerminations := havePendingEarlyTerminations(rt, &st)
+
+	// If we didn't have pending early terminations before, but we do now,
+	// handle them at the next epoch.
+	if !hadEarlyTerminations && hasEarlyTerminations {
+		// First, try to process some of these terminations.
+		if processEarlyTerminations(rt, AddressedSectorsMax) {
+			// If that doesn't work, just defer till the next epoch.
+			scheduleEarlyTerminationWork(rt)
+		}
+		// Note: _don't_ process early terminations if we had a cron
+		// callback already scheduled. In that case, we'll already have
+		// processed AddressedSectorsMax terminations this epoch.
 	}
-	return processMissingPoStFaults(rt, st, store, deadlines, currDeadline.PeriodStart, currDeadline.Index, currDeadline.CurrentEpoch, epochReward, currentTotalPower)
 }
 
-// Discovers how skipped faults declared during post intersect with existing faults and recoveries, then unlocks funds
-// and returns the penalty to be paid and the sector infos needed to deduct power.
-// - Skipped faults that are not in the current deadline will trigger an error.
-// - Skipped faults that have previously been marked recovered, will be penalized as a retracted recovery but will not
-// result in a change in power (the power has already been removed).
-// - Skipped faults that are already declared, but not recovered will be ignored.
-// - The rest will be penalized as undeclared faults and have their power removed.
-func processSkippedFaults(rt Runtime, st *State, store adt.Store, currDeadline *DeadlineInfo, minerInfo *MinerInfo, deadlines *Deadlines, epochReward abi.TokenAmount, currentTotalPower abi.StoragePower, skipped *bitfield.BitField) ([]*SectorOnChainInfo, abi.TokenAmount) {
+// Discovers how skipped faults declared during post intersect with existing faults and recoveries, records the
+// new faults in state.
+// Returns the amount of power newly faulty, or declared recovered but faulty again.
+
+// - Skipped faults that are not in the provided partition triggers an error.
+// - Skipped faults that are already declared (but not delcared recovered) are ignored.
+func processSkippedFaults(rt Runtime, st *State, store adt.Store, faultExpiration abi.ChainEpoch, partition *Partition,
```

Comment:
Can/should we push this down to the state? We'd lose the IllegalArgument/IllegalState distinction, but then the state can maintain relations between the fields.